odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+                    res = native_method(*inputs, **kwargs)
+                except NotImplementedError:
+                    res = super(CupyTensor, self).__array_ufunc__(
+                        ufunc, method, *inputs, **kwargs)
+                else:
+                    if res is NotImplemented:
+                        res = super(CupyTensor, self).__array_ufunc__(
+                            ufunc, method, *inputs, **kwargs)
+
+            else:
+                res = super(CupyTensor, self).__array_ufunc__(
+                    ufunc, method, *inputs, **kwargs)
+
+            # Shortcut for scalar or no return value
+            if np.isscalar(res) or res is None:
+                # The first occurs for `reduce` with all axes,
+                # the second for in-place stuff (`at` currently)
+                return res
+
+            # Wrap result if necessary (lazily)
+            if out is None:
+                if is_floating_dtype(res.dtype):
+                    if res.shape != self.shape:
+                        # Don't propagate weighting if shape changes
+                        weighting = CupyTensorSpaceConstWeighting(1.0,
+                                                                  exponent)
+                    spc_kwargs = {'weighting': weighting}
+                else:
+                    spc_kwargs = {}
+
+                res_space = type(self.space)(
+                    res.shape, res.dtype, self.device, **spc_kwargs)
+                result = res_space.element(res)
+            else:
+                result = out_tuple[0]
+
+            return result
+
+    @property
+    def ufuncs(self):
+        """Access to NumPy style ufuncs.
+
+        Examples
+        --------
+        >>> r2 = odl.rn(2, impl='cupy')
+        >>> x = r2.element([1, -2])
+        >>> x.ufuncs.absolute()
+        rn(2, impl='cupy').element([ 1.,  2.])
+
+        These functions can also be used with broadcasting or
+        array-like input:
+
+        >>> x.ufuncs.add(3)
+        rn(2, impl='cupy').element([ 4.,  1.])
+        >>> x.ufuncs.subtract([3, 3])
+        rn(2, impl='cupy').element([-2., -5.])
+
+        There is also support for various reductions
+        (sum, prod, amin, amax):
+
+        >>> x.ufuncs.sum()
+        -1.0
+        >>> x.ufuncs.prod()
+        -2.0
+
+        They also support an out parameter
+
+        >>> y = r2.element([3, 4])
+        >>> out = r2.element()
+        >>> result = x.ufuncs.add(y, out=out)
+        >>> result
+        rn(2, impl='cupy').element([ 4.,  2.])
+        >>> result is out
+        True
+
+        Notes
+        -----
+        Those ufuncs which are implemented natively on the GPU incur no
+        significant overhead. However, for missing functions, a fallback
+        Numpy implementation is used which causes significant overhead
+        due to data copies between host and device.
+        """
+        # TODO: Test with some native ufuncs, then remove this attribute
```

Comment:
fix this todo

Reply:
This raises a question: If we have a cupy space element and run some numpy code on it that requires a cast to numpy array, should we just go ahead and do it, at the cost of hidden performance loss, or should we raise?
I have a slight preference but want to make sure we're on the same page.

Reply:
Well as always im in the "at least it works" boat. I guess it comes down to how feature complete cupy is, if it has almost everything it is fine, but if you stumble every thing you do, it's gonna be irritating to work with

Reply:
I'm in the same boat, so that's fine then. As far as I can see, they're pretty complete on the ufunc front, but methods like `at` and `accumulated` are largely not implemented. So we need some workaround to use the CUDA code for the obvious ones (`sum`, `prod`, `cumsum`, `cumprod` for `add` and `multiply`), but otherwise drop out to Numpy.

The cupy folks are a bit more worried about efficiency, probably since in neural nets, things are already buried deep down and hard to debug, so this kind of bottleneck danger would be pretty bad for them.

Reply:
Go ahead with that then!

Reply:
I've implemented the manual fiddling.